### ISSUE TITLE
can no longer force a re-download of an image via "kn update" w/o extra stuff

### ISSUE THREAD

**Issue by duglin** [knative/client#1318]
Before v0.21.0 I used to be able to do:
```
$ kn service update echo --image duglin/echo
```
and it would always force a new revision to be created. This was really handy when all I wanted was for the image to be re-downloaded because I did a "docker build" to update it. Notice, I don't have a tag on the image name.

It would be nice if this feature was added back in. W/o it, it's not clear to me how to force this to happen.... at least not w/o some (non-obvious to a newbie) hack, like specifying a revision name or adding `--env foo=$RANDOM` to the cmd line.

**Comment by duglin** [comment]
Playing around with this, it seems like we could restore the old semantics:
- if a revision name is in the ksvc, and one was not specified on the `update`, then clear the current name
- add/update a client specified annotation on the ksvc.spec.annotation with a unique value (e.g. current `generation` value)

That should force serving to create a new revision and avoid the duplicate name issue.

This would actually work for any parameters specified on the cmd line, not just `--image`.  But, as of today, the CLI requires at least one param

**Comment by maximilien** [comment]
> Before v0.21.0 I used to be able to do:
> 
> ```
> $ kn service update echo --image duglin/echo
> ```
> 
> and it would always force a new revision to be created. This was really handy when all I wanted was for the image to be re-downloaded because I did a "docker build" to update it. Notice, I don't have a tag on the image name.

So even if the image has not changed (basically the image bits are the same) you would still want this to create a new revision?

**Comment by maximilien** [comment]
When I try this with a service already created `hello` I get this error:

```bash
➜  client git:(main) ./kn service update hello --image gcr.io/knative-samples/helloworld-go
Error: giving up after 3 retries: admission webhook "validation.webhook.serving.knative.dev" denied the request: validation failed: Saw the following changes without a name change (-old +new): spec.template.metadata.name
{*v1.RevisionTemplateSpec}.ObjectMeta.Annotations["client.knative.dev/user-image"]:
	-: "knativesamples/helloworld"
	+: "gcr.io/knative-samples/helloworld-go"
{*v1.RevisionTemplateSpec}.Spec.PodSpec.Containers[0].Image:
	-: "knativesamples/helloworld"
	+: "gcr.io/knative-samples/helloworld-go"

Run 'kn --help' for usage
```

**Comment by duglin** [comment]
> So even if the image has not changed (basically the image bits are the same) you would still want this to create a new revision?

yes, that was the old semantics

Dunno about your error. Were you on v0.20? You should never see that error when using `kn` since the revName is managed by `kn` (unless the user goes out of their way to touch it).

**Comment by evankanderson** [comment]
It seems like it might be nice to be able to simply run `kn service update echo` without arguments to cause a re-resolution of the docker image tag. I think @navidshaikh is going to write more, but removal of the client-generated names caused this change:

Before:

```diff
  spec:
    template:
      metadata:
-       name: echo-1234
+       name: echo-1235
      spec:
        container:
          image: duglin/echo
```

Note that the image _didn't_ change, only the revision name changed; this was enough to cause server-side generation of a new Revision, which re-resolved the image to an exact tag. We could try to change the server side to do a tag->digest resolution on Services and Configurations, which would cause a server-side update from e.g. `duglin/echo@sha256:.....` to `duglin/echo`, but I'd be concerned about knock-on effects. Note that doing this would have the side effect that `kn service update echo -e NEW=VALUE` _wouldn't_ re-resolve the tag on `duglin/echo`, which would be another behavior change. It would also mean that `kn service update echo` would be a no-op without at least one flag like `--image`.
  


As a reminder, the client-generated names tended to [cause other breakage like this for `kubectl clients`](https://github.com/knative/client/issues/1318#issuecomment-846289618), so I'm not sure we want to roll back the client-side name generation change.

**Comment by duglin** [comment]
I don't think I'm asking for the old client-side name generation. What I'm suggesting is that the CLI ack that the user said "update" and that forces a new revision. I believe this would give us the old semantics at least w.r.t. forcing a re-download. My suggested change [here](https://github.com/knative/client/issues/1318#issuecomment-842591544) should do it. 

Whether we require a param like `--image` or not is secondary to me... I think I'd prefer to have it work w/o any parameters, because I think that presents a nicer UX, but if we required at least one then that's ok - as long as it supports specifying a param value that isn't different from what's there today (which is how I think it used to work, or at least did for --image).

**Comment by maximilien** [comment]
I believe @navidshaikh also suggested a way to have this work with using a revision flag, If I recall correctly. Navid can you please chime in with your suggestion from the call this AM. Thanks.

**Comment by duglin** [comment]
I'd prefer to see if we can keep revision names out of it because, to me, revision names are kind of an advanced topic and probably mainly used when doing traffic splitting. In the simple cases some people will never even really know about revisions. As such, asking them to do something like set a revision name just to force a re-download of the image is asking them to learn about something (to them) unrelated. So to me, this isn't so much about "can they get the net result somehow, it's can they get the net result in an intuitive way". And the old semantics seemed pretty straight-forward to me.

**Comment by maximilien** [comment]
@duglin I also agree that changing semantics for an important command as `service update` is not something we should take lightly or change after it's been consistent from the start. This IMO is a strong argument to not bread old semantics.

**Comment by navidshaikh** [comment]
The use case here is to resolve the image digest, currently thats possible by requesting a new revision and given thats how serving resolves the image digest. Client was doing this by defaulting to client side revision name generation per update. However this caused inconsistencies when users are working with kubectl and kn.

With changing the default of client generating revision name (from v0.21), users can still get the same semantics using
```
kn service update echo --image duglin/echo  --revision-name='{{.Service}}-{{.Random 5}}-{{.Generation}}'
```
where `{{.Service}}-{{.Random 5}}-{{.Generation}}` template fields gets auto-filled for users by kn.

I think this is now more explicit, where client would do as asked, here to create a new revision. This also requires users to know that their image tag will be resolved with new revision generation. I agree this is not very intuitive.


There are other flags in kn, which controls how the tag resolution should happen:
```
      --lock-to-digest                    Keep the running image for the service constant when not
                                          explicitly specifying the image. (--no-lock-to-digest
                                          pulls the image tag afresh with each new revision)
      --no-lock-to-digest                 Do not keep the running image for the service constant
                                          when not explicitly specifying the image.
                                          (--no-lock-to-digest pulls the image tag afresh with
                                          each new revision)
```
this works with client adding/removing an annotation `client.knative.dev/user-image` to ksvc:

https://github.com/knative/client/blob/350b2deffc0415c7d98144e315174f3c1c6d3549/pkg/kn/commands/service/configuration_edit_flags.go#L219-L233

<hr>

should aliasing flag `--revision-name` and its value (as shown above) help? say like `--resolve-image` ? but that would be yet another flag and another implicit operation of new revision creation. 

We discussed about this issue in last client WG call ([notes](https://docs.google.com/document/d/1cD7NkJJhSBpo2Q6RBHrbrSe6R5zjTZgO_YDGAluQ_oI/edit?resourcekey=0-1s0gBQJ2r813ZhTBI8eHdg#)), there was a point that if we should list down all the use cases and see if we need a 'ImageTagResolutionPolicy' on ksvc (like ImagePullPolicy for pod) ? an additional knob that controls how serving should act on image resolution. This way users of kubectl and kn clients can set the knob and have consistent behaviour ?

**Comment by rhuss** [comment]
See also my comments https://github.com/knative/client/issues/398#issuecomment-779654440 and https://github.com/knative/serving/issues/11317#issuecomment-851885865 

tldr;  I'm totally +1 for solving this issue server side, similar to a `ImagePullPolicy` so the (a) other clients can benefit from this and (b) that the use has the choice how image should be updated.

**Comment by duglin** [comment]
I don't have a opinion yet on whether this should be solved via a new server-side config option or not. My bigger concern is that we've changed the semantics of what the 'update' command used to do, and what the user must now do to force a re-download of their image. So, at least for me, I'm viewing this as a breaking change, not just for `kn` but for any other tool that's leveraging the `kn` code under the covers.

I'd like to suggest that we make whatever change is necessary in the CLI/`kn` to restore the old semantics and then we can discuss whether or not to switch over to a server-side solution when/if it's ready.

**Comment by rhuss** [comment]
> My bigger concern is that we've changed the semantics of what the 'update' command used to do, and what the user must now do to force a re-download of their image. So, at least for me, I'm viewing this as a breaking change, not just for kn but for any other tool that's leveraging the kn code under the covers.

Actually we _only_ changed the default, which means as soon as you add the default value yourself like in `kn service create --revision-name {{.Service}}-{{.Random 5}}-{{.Generation}}` then everything should be the same as before. If not it is a bug. That change of the default value has been pointed out in https://github.com/knative/client/releases/tag/v0.21.0 including some reasoning for this change.

**Comment by rhuss** [comment]
If you would have used `--revision-name ""` in `kn service create` you already would have got the current behaviour already.

**Comment by duglin** [comment]
As I mentioned, I don't think we should ask users to have to learn about revisions, or do such a complicated thing such as setting a random revision name, to get the old semantics.

I think the CLI might need to be a bit more careful about breaking backwards compatibility - not only for existing users but for other tooling that is based on it.  Is the CLI going to go v1.0 at the same time as eventing and serving?

**Comment by maximilien** [comment]
I agree with @duglin that "the CLI might need to be a bit more careful about breaking backwards compatibility"

Can we discuss this at the next CLI WG? I'll add an agenda item.

**Comment by rhuss** [comment]
> I think the CLI might need to be a bit more careful about breaking backward compatibility

I agree, but I still don't think that this particular issue is a _backward compatibility_ issue, but revealing a bug of how the client deals with server-side naming of revisions. I.e. in older versions, you would also have an issue with the update of images if you would have used `--revision-name = ""` although the behavior wrt/ to image handling should be exactly the same as without this option. The only thing that is critical to backward compatibility per se is that the default of who is naming the revision has changed. This change has been clearly pointed out in the release notes as well as the reason for this change as there are other serious issues (or better limitations) with the mode of client-named revisions (won't repeat those here). So, if you don't mind about the revision naming details (as @duglin mentions above, rarely somebody cares), then this change should not affect you. The bugs revealed by this change however should be fixed though (not reverting the choice of the default value). We need help from the serving backend for this.

If we naively revert the default value back we still end up with (a) server-side generated names reveal the bug described here still and (b) we suffer from all the other limitations of client-side generated revisions names. This is not a solution imo.

**Comment by duglin** [comment]
Let's separate out the revision naming from the purpose of this issue. While they're related because changing one impacted the other, this issue isn't about revision naming.  This is just about reverting the semantics around when an image will be re-downloaded. How we do that under the covers (for the most part) should be transparent to the user. To me:
```
kn service update NAME --image IMAGE
```
should behave the way it did before, and re-download the image. If we want to expand it to also work in cases w/o the `--image` then that's ok but I think that's a separate discussion. If we want to achieve the desired results via a CLI only change then ok. If we want to do it via a server side change, then ok. However, I think we'd be doing a disservice to our users if we don't revert the "semantics" now (yes meaning a CLI change) while we have discussions with the serving team to decide whether or not there's something that should change in serving - which could take a very long time. Changing to a server side change later on should be transparent to the user.

**Comment by rhuss** [comment]
I hear you and I think we can tackle it purely on the client-side for now, with some improvements if we get server-side support, too.

The idea would be:

* Add an annotation `client.knative.dev/updateTimestamp` that is always set when an update occurs. This will always create a new revision when someone calls kn service update, which is an imperative command to update. A change in the resource is needed somewhere, otherwise, the serving reconciler would never be called.
* We copy over the original image name with its tag to the `image:` field in the container spec. (we have stored this original name in annotation, too). This will force the serving side to re-resolve the image again and uses any newer image with this tag. We won't do that if `--lock-to-digest` is given.

This is how we can trigger the update. We could also just allow a `kn service update myscv` (without any options) that would do exactly this.

As @markusthoemmes convinced me, using an additional 'trigger' annotation is the only way to get into the reconciliation handler. Afair, we have also situations where we do not want a new revision to be created on `kn service update`. In this case, we would just not update the timestamp, but we would need to check the details in the code (don't remember it really).

The remembering of the initial image name and triggering a re-resolve by replacing the digestified image in the container spec with the original image name could be done by the server-side, so that every client would benefit from it and can get a refresh of the image digest.

All of this would still work with server-side revision names so that we wouldn't need to revert the default the revision naming.

@duglin would this work for you ?

**Comment by duglin** [comment]
@rhuss yes I think that would work and supporting it with just `kn service update mysvc` would be nice too. Thanks.

**Comment by rhuss** [comment]
Cool, so let me try to fix it now like suggested. It will be part of the next release. 'happy that we hopefully have now a solution that fits both needs (be able to update an image only and getting rid of client-side revision naming as default).

**Comment by rhuss** [comment]
Unfortunately, the fix did not make it to 0.24.0 as I was able to pick up the issue again only since yesterday. The plan is now to have a quick fix for a 0.24.1 (i.e. adding an update annotation), with a more rigorous fix potentially for 0.25.0

Sorry for the delay, but my calendar exploded in the last six weeks.

**Comment by duglin** [comment]
been there!  thanks for the update
